<a href="https://colab.research.google.com/github/CyAdrienneRamos/DATA100-NB3/blob/main/graded3_pandas_student.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Zomato Exercise

Zomato is a restaurant aggregator website. The platform provides information, menus and user-reviews of restaurants as well as food delivery options from partner restaurants in select cities. You have been provided 3 datasets:

- zomato.csv contains restaurant information like average_cost_for_two, cuisines
- location.csv contains location of the restaurants like address, city, locality & etc
- ratings.csv contains aggregate rating for each restaurant

The different files are connected using the `id` columns which can be found in the different files. The id column is a unique identifier to represent a restaurant. You can use this to **merge** the different files.

In [1]:
import pandas as pd
import numpy as np

1. Merge the zomato, location & ratings file into a single dataframe and store it to the `merged_df` variable **(3pts)**

In [2]:

!git clone https://github.com/CyAdrienneRamos/DATA100-NB3.git

l_df = pd.read_csv("/content/DATA100-NB3/location.csv")
z_df = pd.read_csv("/content/DATA100-NB3/zomato.csv")
r_df = pd.read_csv("/content/DATA100-NB3/ratings.csv")

merged_df = pd.merge(pd.merge(l_df, z_df, how="left"), r_df, how="left")

fatal: destination path 'DATA100-NB3' already exists and is not an empty directory.


In [3]:
assert (6830, 35)==merged_df.shape

cols = ['id', 'address', 'city', 'city_id', 'country_id', 'latitude',
       'locality', 'locality_verbose', 'longitude', 'zipcode',
       'aggregate_rating', 'rating_color', 'rating_text', 'votes',
       'average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view',
       'is_delivering_now', 'is_table_reservation_supported',
       'is_zomato_book_res', 'menu_url', 'mezzo_provider', 'name',
       'photos_url', 'price_range', 'thumb', 'url']

assert sorted(merged_df.columns) == sorted(cols)

2. Create a function that returns the average aggregate rating for each city. Expected returned value is a Series datatype which contains name of city as index and average aggregate rating as value. **(2pts)**

In [4]:
def exercise_2(merged_df):
    return merged_df[["city", "aggregate_rating"]].groupby("city").mean()

In [5]:
res_series = exercise_2(merged_df)
assert True == np.isclose(3.302075, res_series[res_series.index == 'Pasay City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.890741, res_series[res_series.index == 'Valenzuela City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.558333, res_series[res_series.index == 'Malabon City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]

3. Using the apply function of pandas convert the rating_text to a numeric equivalent. You are to introduce a new column in merged_df called `rating_numeric` which is based on the value from rating_text: **(3pts)**

Excellent = 5 <br>
Very Good = 4 <br>
Good = 3 <br>
Average = 2 <br>
Poor = 1 <br>
Not rated = 0 <br>

In [6]:
def convert_rt(val):
    if val=="Excellent":
        return 5
    if val=="Very Good":
        return 4
    if val=="Good":
        return 3
    if val=="Average":
        return 2
    if val=="Poor":
        return 1
    else:
        return 0

merged_df["rating_numeric"] = merged_df["rating_text"].apply(convert_rt)

In [7]:
assert 868 == merged_df[merged_df.rating_numeric == 0].shape[0]
assert True == ('rating_numeric' in merged_df.columns)

4. Create a function which filters the merged dataframe to only contain rows coming in from a specific city. Expected returned value is a dataframe **(2pts)**

In [8]:
def exercise_4(merged_df, city):
    return merged_df.loc[merged_df['city']==city]

In [9]:

assert 'Makati City' == exercise_4(merged_df, 'Makati City').city.unique()[0]
assert (976, 36) == exercise_4(merged_df, 'Makati City').shape

assert 'Mandaluyong City' == exercise_4(merged_df, 'Mandaluyong City').city.unique()[0]
assert (443, 36) == exercise_4(merged_df, 'Mandaluyong City').shape

assert (0, 36) == exercise_4(merged_df, 'Quezon  City').shape

5. Create a function to filter the dataframe to only contain certain columns passed in the function. Expected returned value is a dataframe **(1pt)**

In [10]:
def exercise_5(merged_df, cols_to_filter):
    return merged_df[cols_to_filter]

In [11]:
cols = ['name', 'cuisines', 'city', 'aggregate_rating']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 4) == exercise_5(merged_df, cols).shape

cols = ['average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 11) == exercise_5(merged_df, cols).shape

6.  What is the most common restaurant name in the dataset? **(2pts)** Write down your answer and show code used to come up with the answer.

In [12]:
merged_df.groupby('name').count().nlargest(5,'id')

id  address  city  city_id  country_id  latitude  locality  \
name                                                                          
Starbucks        99       99    99       99          99        99        99   
Jollibee         77       77    77       77          77        77        77   
McDonald's       66       66    66       66          66        66        66   
KFC              54       54    54       54          54        54        54   
BonChon Chicken  49       49    49       49          49        49        49   

                 locality_verbose  longitude  zipcode  ...  mezzo_provider  \
name                                                   ...                   
Starbucks                      99         99        0  ...              99   
Jollibee                       77         77        0  ...              77   
McDonald's                     66         66        0  ...              66   
KFC                            54         54        0  ...              54   
BonChon Chicken                49         49        0  ...              49   

                 photos_url  price_range  thumb  url  aggregate_rating  \
name                                                                     
Starbucks                99           99     99   99                99   
Jollibee                 77           77     77   77                77   
McDonald's               66           66     66   66                66   
KFC                      54           54     54   54                54   
BonChon Chicken          49           49     49   49                49   

                 rating_color  rating_text  votes  rating_numeric  
name                                                               
Starbucks                  99           99     99              99  
Jollibee                   77           77     77              77  
McDonald's                 66           66     66              66  
KFC                        54           54     54              54  
BonChon Chicken            49           49     49              49  

[5 rows x 35 columns]

The most common restaurant name in the dataset is "Starbucks" with 99 entries.

7. Which restaurant has the most expensive average cost for two? **(2pts)** Write down your answer and show code used to come up with the answer.

In [13]:
merged_df[['name', 'average_cost_for_two']].nlargest(5, 'average_cost_for_two')

name  average_cost_for_two
1713                           The Tasting Room                 10000
5      Spiral - Sofitel Philippine Plaza Manila                  6000
1728               Medley Buffet - Okada Manila                  5400
1750  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100
2039  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100

The "The Tasting Room" has the most expensive average cost for two, being 10000.

8. How many restaurants do not have a zipcode? **(1pt)** Write down your answer and show code used to come up with the answer.

In [14]:
merged_df['zipcode'].isna().sum()

6713

6713 of the restaurants do not have a zipcode.

9. How many restaurants in the dataset allow online delivery? **(1pt)** Write down your answer and show code used to come up with the answer.

In [15]:
merged_df['has_online_delivery'].sum()

0

No restaurant offers online delivery

10. Which locality has more expensive food? Poblacion or Salcedo Village? **(3pts)** Write down your answer and show code used to come up with the answer.

In [16]:
merged_df[["locality", "price_range", "average_cost_for_two"]].groupby("locality").describe().loc[['Poblacion',"Salcedo Village"]]

price_range                                               \
                      count      mean       std  min  25%  50%  75%  max   
locality                                                                   
Poblacion              84.0  2.821429  0.661519  1.0  2.0  3.0  3.0  4.0   
Salcedo Village        78.0  2.871795  0.727189  1.0  2.0  3.0  3.0  4.0   

                average_cost_for_two                                        \
                               count        mean         std    min    25%   
locality                                                                     
Poblacion                       84.0  898.214286  417.280815    0.0  600.0   
Salcedo Village                 78.0  989.102564  538.796605  300.0  600.0   

                                        
                   50%     75%     max  
locality                                
Poblacion        800.0  1000.0  2500.0  
Salcedo Village  850.0  1200.0  4000.0

Restaurants in Salcedo Village offer more expensive food as compared to those in Poblacion.